### PR TITLE
Return multierror from glooctl check resources

### PR DESCRIPTION
**Pull request by saiskee**
# Description

Please include a summary of the changes.

This fixes the bug where `glooctl check` exits after one failed check. Errors from all checks are now aggregated and displayed after all checks have been completed.

# Context

Users ran into this bug running `glooctl check` and the program exited after one failed check. Completing all checks and reporting all errors was more desirable.

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/2952

### CODE REVIEW

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 1]
```diff
@@ -70,92 +71,106 @@ func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.
...
 
-func CheckResources(opts *options.Options) (bool, error) {
+func CheckResources(opts *options.Options) (bool, *multierror.Error) {
+	var multiErr *multierror.Error
+	var status bool
+
 	err := checkConnection(opts.Metadata.Namespace)
 	if err != nil {
-		return false, err
+		multiErr = multierror.Append(multiErr, err)
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go:80]: this is the one case where we probably don't want to append to the multierror and just want to return early. if the connection to kubernetes is botched, none of the following checks will work -- and the error messages from those will just be noise

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 2]
```diff
@@ -70,92 +71,106 @@ func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.
 	return cmd
 }
 
-func CheckResources(opts *options.Options) (bool, error) {
+func CheckResources(opts *options.Options) (bool, *multierror.Error) {
+	var multiErr *multierror.Error
+	var status bool
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go:76]: realized this while reviewing the pr, we should be able to remove the `bool` return value and just use the multierror.

implicit in a non-nil error is that `ok/status` would be false, so no need to return two values that signal the same thing

**Review by kdorosh — COMMENTED** [review]
looks good so far! two nits and then we can work on some tests 🎉

File: projects/gloo/cli/pkg/cmd/check/root.go (hunk 2 above)
Comment (saiskee) [reply projects/gloo/cli/pkg/cmd/check/root.go:76]: Hmm so the only reason why I included this status boolean variable is because some checks like `getAndCheckDeployments (root.go:190)` return `false` for the boolean and `nil` for the error. This means we may encounter the case where `ok` is false but there is no error.

Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/check/root.go:76]: nice catch!

the only reason we need to do that is because of this line https://github.com/solo-io/gloo/blob/40daeed2cafa51a795a94f7de38d8cdf7500cb09/projects/gloo/cli/pkg/cmd/check/root.go#L175 

i think it would be nicer to change the signature of the function to return (`*appsv1.DeploymentList, error`) and change the line above to:
```golang
		fmt.Printf("Gloo is not installed\n")
		return nil, errors.New("Gloo is not installed")
```

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 3]
```diff
@@ -258,36 +251,36 @@ func checkPods(opts *options.Options) (bool, error) {
...
 				}
 			default:
 				fmt.Printf("Note: Unhandled pod condition %s", condition.Type)
 			}
 
 			if errorToPrint != "" {
-				fmt.Print(errorToPrint)
-				return false, err
+				fmt.Println(errorToPrint)
+				return fmt.Errorf(errorToPrint)
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go:278]: i think we should be returning a multierror here -- if multiple pods aren't ready it would be nice to report them all

swapping `errorToPrint` for a multierror seems like the change we want here

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 4]
```diff
@@ -191,36 +194,26 @@ func getAndCheckDeployments(opts *options.Options) (*appsv1.DeploymentList, bool
...
+				return nil, err
 			}
 		}
 
 		for _, condition := range deployment.Status.Conditions {
 			setMessage(condition)
 			if condition.Type == appsv1.DeploymentAvailable && condition.Status != corev1.ConditionTrue {
-				errorToPrint = fmt.Sprintf("Deployment %s in namespace %s is not available!%s\n", deployment.Name, deployment.Namespace, message)
+				err := fmt.Errorf("Deployment %s in namespace %s is not available!%s\n", deployment.Name, deployment.Namespace, message)
+				return nil, err
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go:214]: same here, reporting all deployments here with multierror would be nice

File: projects/gloo/cli/pkg/cmd/check/root_test.go [hunk 5]
```diff
@@ -62,6 +63,78 @@ var _ = Describe("Root", func() {
...
+				Metadata: core.Metadata{
+					Name:      "some-warning-upstream",
+					Namespace: "gloo-system",
+				},
+				Status: core.Status{
+					State:  core.Status_Warning,
+					Reason: "I am an upstream with a warning",
+				},
+			},
+				clients.WriteOpts{})
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root_test.go:102]: super nit, this indentation is a bit off, we usually but these on one line

File: test/kube2e/gloo-mtls/gloo_mtls_suite_test.go [hunk 6]
```diff
@@ -72,14 +72,11 @@ var _ = BeforeSuite(func() {
 				Namespace: testHelper.InstallNamespace,
 			},
 		}
-		ok, err := check.CheckResources(opts)
+		err := check.CheckResources(opts)
 		if err != nil {
-			return errors.Wrapf(err, "unable to run glooctl check")
+			return errors.New("glooctl check detected a problem with the installation")
```
Comment (kdorosh) [inline test/kube2e/gloo-mtls/gloo_mtls_suite_test.go:77]: why this change? i think wrapping the error is desirable here

**Review by kdorosh — COMMENTED** [review]
overall looks good, very nice test! just a couple nits

File: projects/gloo/cli/pkg/cmd/check/root_test.go (hunk 5 above)
Comment (saiskee) [reply projects/gloo/cli/pkg/cmd/check/root_test.go:102]: So gofmt doesn't break up long lines, so I tried using another formatter called golines (https://github.com/segmentio/golines). I think it's still not refined, so I stopped using it, but that's probably the reason for this.

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 7]
```diff
@@ -220,10 +220,15 @@ func getAndCheckDeployments(opts *options.Options) (*appsv1.DeploymentList, erro
...
 				condition.Type != appsv1.DeploymentReplicaFailure &&
 				condition.Type != appsv1.DeploymentProgressing {
-				fmt.Printf("Note: Unhandled deployment condition %s", condition.Type)
+				err := fmt.Errorf("Deployment %s has an unhandled deployment condition %s", deployment.Name, condition.Type)
+				multiErr = multierror.Append(multiErr, err)
 			}
 		}
 	}
+	if multiErr != nil {
+		fmt.Printf("%v Errors!\n", multiErr)
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/check/root.go:229]: multierror.len()?

**Review by kdorosh — APPROVED** [review]
🚀
